nemo/collections/nlp/modules/common/megatron/retrieval_service.py

```diff
@@ -13,64 +13,174 @@
 # limitations under the License.
 
 import abc
+import base64
+import json
+import logging
+import pickle
+import threading
 import time
 from typing import List, Union
 
 import faiss
 import numpy as np
+import requests
 import torch
+from flask import Flask, jsonify, request
+from flask_restful import Api, Resource
 from sentence_transformers import SentenceTransformer
 
 from nemo.collections.common.tokenizers.tokenizer_spec import TokenizerSpec
 from nemo.collections.nlp.data.language_modeling.megatron.indexed_retrieval_dataset import MMapRetrievalIndexedDataset
-from nemo.utils import logging
+
+log = logging.getLogger('werkzeug')
+log.setLevel(logging.ERROR)
+
+lock = threading.Lock()
+headers = {"Content-Type": "application/json"}
+
+PORT_NUM = 17179
+PORT_NUM_DYN = 17180
+PORT_NUM_BERT = 17181
+
+
+def request_data(data, port=PORT_NUM):
+    resp = requests.put('http://localhost:{}/knn'.format(port), data=json.dumps(data), headers=headers)
+    return resp.json()
 
 
 class RetrievalService:
     @abc.abstractmethod
-    def get_knn(self, query: Union[List[str], str, torch.Tensor]):
+    def get_knn(self, query: Union[List[str], str, torch.Tensor], neighbors: int):
         pass
 
+    @abc.abstractmethod
+    def add_docs_to_index(self, docs: List[str], add_eos: bool = True):
+        """
+        Add documents to the Faiss index
+        Args:
+            docs: List[str], list of documents that is going to be added to the index
+            add_eos: bool, whether add the eos in the end
+        """
+        raise NotImplementedError()
 
-class FaissRetrievalService(RetrievalService):
+
+class ChunkStore:
+    def __init__(self, chunk_size, pad_id):
+        self.store = {}
+        self._count = 0
+        self.no_retrieval = np.ones(2 * chunk_size, dtype=np.int64) * pad_id
+        self.store[-1] = self.no_retrieval
+
+    def add(self, chunk):
+        self.store[self._count] = chunk
+        self._count += 1
+
+    def get_chunk(self, neighbor_id):
+        return self.store[neighbor_id]
+
+    def reset(self):
+        self._count = 0
+        self.store = {}
+        self.store[-1] = self.no_retrieval
+
+
+class SentenceBertResource(Resource):
+    def __init__(
+        self, bert_model, tokenizer, pool, sentence_bert_batch,
+    ):
+        # server
+        self.bert_model = bert_model
+        self.tokenizer = tokenizer
+        self.pool = pool
+        self.sentence_bert_batch = sentence_bert_batch
+        self.embedding_dim = self.bert_model.get_sentence_embedding_dimension()
+
+    def put(self):
+        # logging.info("request IP: " + str(request.remote_addr))
+        # logging.info(json.dumps(request.get_json()))
+        data = request.get_json()
+        if isinstance(data, dict):
```

Comment:
Why do you have this check here? The same doesn't exist in `FaissRetrievalResource`